gateway-api/src/gateway_api/time_duration.rs

```diff
@@ -0,0 +1,315 @@
+        Self::new(secs, 0)
+    }
+
+    /// gateway_api::Duration::from_micros creates a new gateway_api::Duration
+    /// from microseconds, but requires that the resulting Duration be valid
+    /// according to GEP-2257.
+    pub fn from_micros(micros: u64) -> Result<Self, String> {
+        let sec = micros / 1_000_000;
+        let ns = ((micros % 1_000_000) * 1_000) as u32;
+
+        Self::new(sec, ns)
+    }
+
+    /// gateway_api::Duration::from_millis creates a new gateway_api::Duration
+    /// from milliseconds, but requires that the resulting Duration be valid
+    /// according to GEP-2257.
+    pub fn from_millis(millis: u64) -> Result<Self, String> {
+        let sec = millis / 1_000;
+        let ns = ((millis % 1_000) * 1_000_000) as u32;
+
+        Self::new(sec, ns)
+    }
+
+    delegate! {
+        to self.0 {
+            /// gateway_api::Duration::as_secs returns the number of whole
+            /// seconds in the duration.
+            pub fn as_secs(&self) -> u64;
+
+            /// gateway_api::Duration::as_millis returns the number of whole
+            /// milliseconds in the duration.
+            pub fn as_millis(&self) -> u128;
+
+            /// gateway_api::Duration::as_micros returns the number of whole
+            /// microseconds in the duration.
+            pub fn as_nanos(&self) -> u128;
+
+            /// gateway_api::Duration::subsec_nanos returns the number of
+            /// nanoseconds in the duration that are not part of the whole
+            /// seconds.
+            pub fn subsec_nanos(&self) -> u32;
+
+            /// gateway_api::Duration::is_zero returns true if the duration is
+            /// zero.
+            pub fn is_zero(&self) -> bool;
+        }
+    }
+}
+
+impl FromStr for Duration {
+    type Err = String;
+
+    /// Parsing a gateway_api::Duration from a string requires that the string
+    /// match the GEP-2257 duration format, and that the resulting duration be
+    /// valid according to GEP-2257.
+    fn from_str(duration_str: &str) -> Result<Self, Self::Err> {
+        // GEP-2257 dictates that string values must match this regex and be
+        // parsed the same way that Go's time.ParseDuration parses durations.
+        let re = Regex::new(r"^([0-9]{1,5}(h|m|s|ms)){1,4}$").unwrap();
+
+        // If the string doesn't match the regex, it's invalid.
+        if ! re.is_match(duration_str) {
+            return Err("Invalid duration format".to_string());
+        }
+
+        // We use kube::core::Duration to do the heavy lifting of parsing.
+        let kd = k8sDuration::from_str(duration_str);
+
+        // If the parse fails, return an error immediately...
+        if kd.is_err() {
+            return Err(kd.err().unwrap().to_string());
+        }
+
+        // ...otherwise, we need to check the duration for validity by turning
+        // it into a gateway_api::Duration.
+        Duration::try_from(stdDuration::from(kd.unwrap()))
```

Comment:
```suggestion
        // We use kube::core::Duration to do the heavy lifting of parsing.
        match k8sDuration::from_str(duration_str) {
            // If the parse fails, return an error immediately...
            Err(error) => Err(error.to_string()),
            // ...otherwise, we need to check the duration for validity
            // by turning it into a gateway_api::Duration.
            Some(duration) =>  Duration::try_from(stdDuration::from(duration))
        }   
```